File: frontend/src/types/encounter.ts
export interface EncounterGenerationRequest {
    character_levels: number[];
    difficulty: 'easy' | 'medium' | 'hard' | 'deadly';
    monsters: Record<string, number>;
}

export interface EncounterMonster {
    name: string;
    initiative: number;
    hp: number;
}

export interface EncounterGenerationResult {
    monsters: EncounterMonster[];
    monster_xp_total: number;
    monster_xp_with_modifiers: number;
    difficulty_rating: string;
    party_difficulty_thresholds: Record<string, number>;
    player_count: number;
}

export interface XpLevels {
    easy: number;
    medium: number;
    hard: number;
    deadly: number;
}

// Used by a user to build an encounter
export interface EncounterBuilder {
    monsters: Record<string, number>; // monster name to quantity
}

File: backend/app/models/encounter.py
"""
This module contains the encounter models.
"""
from pydantic import field_validator

from ..core.model_base import ModelBase
from ..core.types import EncounterDifficulty


class EncounterGenerationRequest(ModelBase):
    character_levels: list[int]
    difficulty: EncounterDifficulty
    monsters: dict[str, int]  # Monster name to max occurrence

    @field_validator('difficulty', mode='before')
    def normalize_size(cls, value):
        if isinstance(value, str):
            formatted_value = value.replace('_', ' ').replace('-', ' ').strip().lower()
            for e in EncounterDifficulty:
                if e.value.lower() == formatted_value or e.name.lower() == formatted_value:
                    return e
        return value


class EncounterMonster(ModelBase):
    name: str
    initiative: int = 1
    hp: int = 1


class EncounterGenerationResult(ModelBase):
    monsters: list[EncounterMonster]
    monster_xp_total: int
    monster_xp_with_modifiers: int
    difficulty_rating: EncounterDifficulty
    party_difficulty_thresholds: dict[str, int]
